### PR TITLE
Regenerate health in darkness flag

### PR DESCRIPTION
**Pull request by John-Candlebury**
This pr adds a flag that allows enemies to quickly regenerate health in the dark. Gave the flag to the wraith for now, so make sure to get flashlights.

### CODE REVIEW

**Comment by kevingranade** [comment]
Nice idea!

**Comment by Coolthulhu** [comment]
If they can be defeated by a mere flashlight, it isn't really spooky.
How about scaling it with light and not just a threshold? So that bigger flashlight weakens them more and only floodlight completely shuts them down.
You can get actual light level with `map::ambient_light_at`. Sunlight is either 60 or 100 (I'm not sure which, but debug menu map editor shows light on a tile), so 30-50 should be a good point to stop regenerating.

As far as I recall, we have a policy of not accepting untestable content. So this PR should be a part of the one that adds the monster that makes the use of the flag.

**Comment by John-Candlebury** [comment]
>You can get actual light level with map::ambient_light_at. Sunlight is either 60 or 100 (I'm not sure which, but debug menu map editor shows light on a tile), so 30-50 should be a good point to stop regenerating.

Made the heal scale down linearly (I dont think more complex scaling is worth it) with ambient light. It doesn't seem to have much of an effect, since flashlights and other lighting items create light sources with intensity in the hundreds, so as long as you keep close the enemies won't get regeneration.

Maybe we could make the threshold for what qualifies as a light tile a bit higher, to around 50 ambient light instead of 10? But I am unsure of how items like the atomic lamps would fare then.

>As far as I recall, we have a policy of not accepting untestable content. So this PR should be a part of the one that adds the monster that makes the use of the flag.

I do meant for the wraith to have the flag,and for a few other monsters to have it as well, so its not an provisional change.

File: src/monster.cpp [hunk 1]
```diff
@@ -1740,6 +1740,10 @@ void monster::process_effects()
         add_msg( m_warning, _( "The %s seems a little healthier." ), name().c_str() );
     }
 
+    if( has_flag( MF_REGENERATES_IN_DARK ) && heal( 40 - int( g->m.ambient_light_at( pos() ) ) ) > 0 && one_in( 2 ) && g->u.sees( *this ) ) {
+        add_msg( m_warning, _( "The %s uses the darkness to regenerate." ), name().c_str() );
```
Comment (keyspace) [inline src/monster.cpp]: I'd avoid specifying that it "uses darkness". The flag may be useful for other monsters that regenerate for a different reason.

How about the same generic `_( "The %s is visibly regenerating!" )` as above? That'd even save a line for translators.

File: data/json/overmap_terrain.json [hunk 2]
```diff
@@ -1602,6 +1602,15 @@
         "known_down" : true
     },{
         "type" : "overmap_terrain",
+        "id" : "standing_stones",
+        "name" : "forest?",
```
Comment (keyspace) [inline data/json/overmap_terrain.json]: There's a `cabin_strange` that doesn't bother with a '?' hint. Maybe this one shouldn't either?..

File: data/json/mapgen/standing_stones.json [hunk 3]
```diff
@@ -0,0 +1,59 @@
+[
+  {
+    "name": "GROUP_STANDING_STONES",
+    "type": "monstergroup",
+    "default": "mon_null",
+    "monsters": [
+      { "monster": "mon_darkman", "freq": 100, "cost_multiplier": 1, "conditions" : ["NIGHT"] }
```
Comment (John-Candlebury) [inline data/json/mapgen/standing_stones.json]: There appears to be an issue with the ` "conditions" : ["NIGHT"]` but there is no loading error when starting a game and the syntax is the same as the one used in monstergroups.json.

Could this be an issue with the linter?

File: src/monster.cpp (hunk 1 above)
Comment (John-Candlebury) [inline src/monster.cpp]: If the darkness part is removed, how is a player going to know that a lightsource can prevent the regeneration?

File: data/json/mapgen/standing_stones.json [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        "s": "t_dirt",
+        "o": "t_dirt",
+        "O": "t_dirt"
+      },
+      "furniture": {
+        "o": "f_boulder_small",
+        "O": "f_boulder_large"
+      },
+      "place_loot": [
+        { "item": "stick", "x": 10, "y": 12, "repeat": 2 }
```
Comment (keyspace) [inline data/json/mapgen/standing_stones.json]: I'll submit a PR to your repo/branch showing how this can be mappable (just as furniture above).

File: data/json/mapgen/standing_stones.json (hunk 3 above)
Comment (keyspace) [inline data/json/mapgen/standing_stones.json]: Yes, will submit a separate PR.

**Comment by keyspace** [comment]
Blocked until PR #20064 or similar is merged.

File: src/monster.cpp (hunk 1 above)
Comment (keyspace) [inline src/monster.cpp]: It's very obvious now. `The %s gets darker and regenerates.`?.. Honestly, I don't know. :/

Comment (John-Candlebury) [inline src/monster.cpp]: What about: `The %s appears healthier in the darkness!`  ?

Comment (keyspace) [inline src/monster.cpp]: Could work! I mean, the player character sees something happening, but doesn't necessarily know the mechanism by which it happens. It's not the "darkness" that irks me, it's "uses".

**Comment by keyspace** [comment]
Please add the flag to `doc/JSON_FLAGS.md` (search `REGENERATES` for location).

**Comment by Coolthulhu** [comment]
> It doesn't seem to have much of an effect, since flashlights and other lighting items create light sources with intensity in the hundreds, so as long as you keep close the enemies won't get regeneration.

Are the creatures supposed to survive sunlight? If not, you could bump the scaling number to heavy-duty flashlight level, so that regular one only slows it down.
If yes, special case "if not in sunlight" to make sunlight cancel all such regeneration no matter how much sun is there.

> Maybe we could make the threshold for what qualifies as a light tile a bit higher, to around 50 ambient light instead of 10?

I'd be wary of changing global, low-level light constants without good testing.
Would probably be easier to allow sunlight to reach thousands of light units. Most sunlight-dependent calculations, such as solars, use the maximum sunlight level to determine power, so they shouldn't break.

This would have a neat effect of allowing effects like glare to proc off regular light if it was intense enough instead of having it hardcoded as sunlight, as if player was some sort of a vampire.

File: src/monster.cpp [hunk 5]
```diff
@@ -1740,8 +1740,11 @@ void monster::process_effects()
         add_msg( m_warning, _( "The %s seems a little healthier." ), name().c_str() );
     }
 
-    if( has_flag( MF_REGENERATES_IN_DARK ) && heal( 40 - int( g->m.ambient_light_at( pos() ) ) ) > 0 && one_in( 2 ) && g->u.sees( *this ) ) {
+    if( has_flag( MF_REGENERATES_IN_DARK ) ) {
+        float light = g->m.ambient_light_at( pos() );
```
Comment (keyspace) [inline src/monster.cpp]: Can be a `const`?

**Comment by John-Candlebury** [comment]
New scaling works  better, flashlights prevent regent only when adjacent, heavy flashlights at closer than 3 tiles, and a floodlight prevents regen at closer than 20 tiles.

**Comment by keyspace** [comment]
`2.71` - magic number. Euler's? Mention in a comment. There's only one other place in the codebase that uses it (as `2.7`), so not a `constexpr` yet.

Alternatively, `std::exp(1.0)`?

~~`1000`~~ `10000` also magic. Was it chosen from possible current values of `g->m.ambient_light_at()`? If so, and the scaling to that were to ever change, this might get broken. Anyway, mention in a comment, too.

File: src/monster.cpp [hunk 6]
```diff
@@ -1740,6 +1740,13 @@ void monster::process_effects()
         add_msg( m_warning, _( "The %s seems a little healthier." ), name().c_str() );
     }
 
+    if( has_flag( MF_REGENERATES_IN_DARK ) ) {
+        const float light = g->m.ambient_light_at( pos() );
+        if ( heal ( int( 50.0 *  pow( 2.71 , - light*light / 10000 ) )  > 0 && one_in( 2 ) && g->u.sees( *this ) ) ) {
```
Comment (Coolthulhu) [inline src/monster.cpp]: `pow( 2.71, x )` is more properly expressed as `exp( x )`.
